packages/stylist-core/src/parser.rs

```diff
@@ -20,46 +20,80 @@ use nom::{
 #[cfg(test)]
 use log::trace;
 
-/// A lightweight CSS Parser.
-#[derive(Debug)]
-pub(crate) struct Parser {}
-
+/// Wrap a parser, tracing input and output.
+// if not cfg(test), this would trip up clippy.
 #[allow(clippy::let_and_return)]
-impl Parser {
-    /// Returns Error when string is Empty
-    fn expect_non_empty(i: &str) -> std::result::Result<(), nom::Err<VerboseError<&str>>> {
+fn traced_context<I, O, F>(
+    ctx: &'static str,
+    mut p: impl nom::Parser<I, O, F>,
+) -> impl FnMut(I) -> IResult<I, O, F>
+where
+    I: fmt::Display + fmt::Debug + Clone,
+    O: fmt::Debug,
+    F: fmt::Debug + nom::error::ContextError<I>,
+{
+    move |i| {
+        #[cfg(test)]
+        trace!("> {}: {}", ctx, i);
+
+        let result = context(ctx, |i| p.parse(i))(i);
+
+        #[cfg(test)]
+        trace!("< {}: {:#?}", ctx, result);
+
+        result
+    }
+}
+
+/// Returns Error when string is Empty
+fn expect_non_empty<'a, O, F>(
+    mut p: impl nom::Parser<&'a str, O, F>,
+) -> impl FnMut(&'a str) -> IResult<&'a str, O, F>
+where
+    F: nom::error::ParseError<&'a str>,
+{
+    move |i| {
         if i.is_empty() {
             Err(nom::Err::Error(ParseError::from_error_kind(
                 i,
                 ErrorKind::LengthValue,
             )))
         } else {
-            Ok(())
+            p.parse(i)
         }
     }
+}
+
+/// A lightweight CSS Parser.
+#[derive(Debug)]
+pub(crate) struct Parser {}
 
+// Naming convention:
+// - parsers ending with _ns assume the input has *no space* (superfluous whitespace) at the start
```

Comment:
I am about to add rules in rules and rules in blocks and this behaviour is not desirable when a parser function is used in multiple places.

IMO, The parser function should always try to strip spaces.

So developers do not have to follow naming convention to make sure that the parser is functional.

Reply:
In addition, this convention is not consistent. For example: `ident` parser does not accept any spaces, but does not have `_ns` affix.

Reply:
It's trying to avoid trimming multiple times already knowing the input can't be trimmed. I'm wondering, maybe it's better to put that part in a separate PR.

Reply:
The current parser performance is OK to me?

If parser is used at compile time, performance does not matter that much as long as the performance is not horrible.
If parser is used at runtime, IMO, file size is more important than performance as long as the performance is not horrible.

I was actually wondering if we should continue using nom after inspecting the parser used by `emotion` and its code size.

https://github.com/thysultan/stylis.js/blob/master/src/Parser.js

Reply:
Okay, let me convert this PR back into draft and shrink the changes down even more to address specfically the linked bug and improved debugging with `traced_context` only, the rest in another issue, sounds good?

Reply:
Sounds good to me.